dali/pipeline/executor/executor2/exec_graph_analysis.cc

```diff
@@ -0,0 +1,282 @@
+        auto *inp = node->inputs[i];
+        inp->producer->outputs[inp->producer_output_idx].pinned = true;
+      }
+    } else if (node->op->GetSpec().GetSchema().HasPassThrough()) {
+      auto &schema = node->op->GetSpec().GetSchema();
+      int nout = node->outputs.size();
+      for (int i = 0; i < ninp; i++) {
+        auto *input = node->inputs[i];
+        if (input->device != StorageDevice::CPU)  // we're not interested in non-CPU buffers
+          continue;
+
+        auto &source_output = input->producer->outputs[input->producer_output_idx];
+        if (source_output.pinned)  // already pinned
+          continue;
+
+        for (int o = 0; o < nout; o++) {
+          // If input `i` passes to a pinned output `o`, then the input should also be marked
+          // as pinned. This will be followed in reverse topological order.
+          if (node->outputs[o].pinned && schema.IsPassThrough(i, o, false)) {
+            source_output.pinned = true;
+            break;
+          }
+        }
+      }
+    }
+  }
+};
+
+class ExecGraph::SortHelper {
+ public:
+  explicit SortHelper(ExecGraph &graph) : graph_(graph) {}
+
+  using NodeList = std::list<ExecNode>;
+  using NodeIt = NodeList::iterator;
+
+  void Run() {
+    sorted_.reserve(graph_.nodes_.size());
+
+    graph::ClearVisitMarkers(graph_.nodes_);
+
+    // Stable topological sort - if the graph is already sorted, then the order shouldn't change
+    for (auto &node : graph_.nodes_)
+      SortNode(&node);
+
+    MoveBatchSizeProvidersToFront();
+    RecreateNodeList();
+  }
+
+  void SortNode(ExecNode *node) {
+    assert(node);
+    graph::Visit visit(node);
+    if (!visit)
+      return;
+    for (auto *edge : node->inputs) {
+      assert(edge);
+      SortNode(edge->producer);
+    }
+    int idx = sorted_.size();
+    sorted_.push_back(node);
+  }
+
+  void MoveBatchSizeProvidersToFront() {
+    std::stable_partition(sorted_.begin(), sorted_.end(), [](const ExecNode *n) {
+      return n->is_batch_size_provider;
+    });
+  }
+
+  void RecreateNodeList() {
+    std::unordered_map<const ExecNode *, NodeIt> node2it;
+    for (auto it = graph_.nodes_.begin(); it != graph_.nodes_.end(); ++it)
+      node2it[&*it] = it;
+
+    NodeList sorted_list;
+    for (auto *node : sorted_)
+      sorted_list.splice(sorted_list.end(), graph_.nodes_, node2it[node]);
+
+    assert(graph_.nodes_.empty() && "Everything should have been moved to the sorted list");
+    graph_.nodes_ = std::move(sorted_list);
+  }
+
+  ExecGraph &graph_;
+  std::vector<ExecNode *> sorted_;
+};
+
+void ExecGraph::Sort() {
+  if (sorted_)
+    return;
+
+  SortHelper sort(*this);
+  sort.Run();
+  sorted_ = true;
+}
+
+void ExecGraph::Analyze() {
+  if (analyzed_)
+    return;
+  Analyzer a;
+  a.FindPinnedBuffers(*this);
+  a.MarkOutputsWithParallelConsumers(*this);
+  analyzed_ = true;
+}
+
+void ExecGraph::Validate() {
+  // The checks here are extremely defensive, but they're only run once.
```

Comment:
This is a bug-check, as clearly stated by the "Internal error" prefix. I decided to go with that instead of asserts to (extremely defensively) guard against attempting to run a structurally malformed graph. The aim here is to avoid very hard to diagnose errors, like dangling node/edge pointers that are valid addresses but invalid from the graph structure standpoint.